File: similo-thesis/src/main/scala/elements/Element.scala
package elements

import io.circe.generic.auto.*
import io.circe.parser.*
import io.circe.syntax.*
import legacy.LegacyElementUtils
import org.openqa.selenium.WebElement
import org.openqa.selenium.remote.RemoteWebDriver
import util.Utils.*

import java.awt.Rectangle
import scala.jdk.CollectionConverters.*

case class Element private (
  tag: String,
  innerText: Option[String],
  nodeText: Option[String],
  childrenLength: Int,
  neighbourText: Option[String] = None,
  xpath: Xpath,
  cssSelector: String,
  location: (Int, Int),
  dimension: (Int, Int),
  attributes: Map[String, String],
  cssStyles: Map[String, String]
):

  def this(
    tag: String,
    innerText: Option[String],
    nodeText: Option[String],
    childrenLength: Int,
    xpath: Xpath,
    cssSelector: String,
    location: (Int, Int),
    dimension: (Int, Int),
    attributes: Map[String, String],
    cssStyles: Map[String, String]
  ) = this(
    tag,
    innerText,
    nodeText,
    childrenLength,
    None,
    xpath,
    cssSelector,
    location,
    dimension,
    attributes.map((k, v) => (k, removeWayBackUrls(v))),
    cssStyles.map((k, v) => (k, removeWayBackUrls(v)))
  )

  override def equals(obj: Any): Boolean = obj match
    case that: Element =>
      tag == that.tag &&
      innerText == that.innerText &&
      nodeText == that.nodeText &&
      childrenLength == that.childrenLength &&
      neighbourText == that.neighbourText &&
      xpath == that.xpath &&
      cssSelector == that.cssSelector &&
      location == that.location &&
      dimension == that.dimension &&
      attributes == that.attributes
    case _ => false

  def name: Option[String] = attributes.get("name")

  def id: Option[String] = attributes.get("id")

  def href: Option[String] = attributes.get("href")

  def title: Option[String] = attributes.get("title")

  def alt: Option[String] = attributes.get("alt")

  def area: Int = width * height

  def height: Int = dimension._2

  def width: Int = dimension._1

  def shape: Int = if height == 0 then 0 else (width * 100) / height

  def ariaLabel: Option[String] = attributes.get("aria-label")

  def isButton: Boolean =
    if tag == null then false
    else if tag.is("a") && className.indexOf("btn") >= 0 then true
    else if tag.is("button") then true
    else tag.is("input") && (typeName.is("button") || typeName.is("submit") || typeName.is("reset"))

  def className: Option[String] = attributes.get("class")

  def typeName: Option[String] = attributes.get("type")

  def addNeighborTexts(string: String): Element =
    copy(neighbourText = Some(string))

  def addNeighborTexts(strings: Seq[Element]): Element =
    LegacyElementUtils.addNeighborTexts(this, strings.asJava)

  def getVisibleText: Option[String] = Seq(innerText, value, placeholder).flatten.find(_.trim.nonEmpty)

  def value: Option[String] = attributes.get("value")

  def placeholder: Option[String] = attributes.get("placeholder")

  def getLocationArea: Rectangle = new Rectangle(x, y, width, height)

  def isNotEmpty: Boolean = !isEmpty

  def isEmpty: Boolean = xpath.chrome == "-"

  def json: String = this.asJson.spaces2

  def tolerantEquals(that: Element): Boolean =
    tag == that.tag &&
      nodeText == that.nodeText &&
      Math.abs(x - that.x) < 10 &&
      Math.abs(y - that.y) < 10 &&
      Math.abs(width - that.width) < 5 &&
      Math.abs(height - that.height) < 5 &&
      attributes == that.attributes

  def y: Int = location._2

  def x: Int = location._1

object Element:

  private final val attributeExtractorScript = ressource("javascript/attribute_extractor.js").get
  private final val cssAttributesExtractorScript = ressource("javascript/css_attribute_extractor.js").get
  private final val xPathExtractorScript = ressource("javascript/xpath_extractor.js").get
  private final val cssSelectorExtractorScript = ressource("javascript/css_selector_extractor.js").get
  private final val directTextExtractorScript = ressource("javascript/direct_text_extractor.js").get
  private final val elementsExtractorScript = ressource("javascript/elements_extractor.js").get

  private final val scriptsCombined = s"""
       |$attributeExtractorScript
       |$cssAttributesExtractorScript
       |$xPathExtractorScript
       |$cssSelectorExtractorScript
       |$directTextExtractorScript
       |$elementsExtractorScript
       |""".stripMargin

  private final val elementsExtractor = s"""
       |$scriptsCombined
       |return elements_extractor();
       |""".stripMargin

  private final val elementExtractor = s"""
        |$scriptsCombined
        |return element_extractor_string(arguments[0]);
        |""".stripMargin

  def extractAll(driver: RemoteWebDriver): Seq[Element] =
    val elementsString = driver.executeScript(elementsExtractor).toString
    parseElements(elementsString)

  def parseElements(string: String): Seq[Element] =
    decode[Seq[Element]](string).getOrElse(throw new Exception(s"Could not parse elements: $string"))

  def extract(driver: RemoteWebDriver, element: WebElement): Element =
    val elementsString = driver.executeScript(elementExtractor, element).toString
    parseElement(elementsString)

  private def parseElement(string: String): Element =
    decode[Element](string).getOrElse(throw new Exception(s"Could not parse element: $string"))

  def empty: Element = Element(
    tag = "",
    nodeText = None,
    innerText = None,
    childrenLength = 0,
    xpath = Xpath.empty,
    cssSelector = "",
    location = (0, 0),
    dimension = (0, 0),
    attributes = Map.empty,
    cssStyles = Map.empty
  )

  def justTag(tag: String): Element = Element(
    tag = tag,
    nodeText = None,
    innerText = None,
    childrenLength = -1,
    xpath = Xpath.empty,
    cssSelector = "",
    location = (-1, -1),
    dimension = (-1, -1),
    attributes = Map.empty,
    cssStyles = Map.empty
  )

File: similocator/src/main/java/de/tum/ls4/artifacts/Element.java
package de.tum.ls4.artifacts;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Stream;

import static de.tum.ls4.utils.Utils.checkedCastToMap;

public class Element implements Artifact {
    // Basic Properties
    public final String tag;
    public final String innerText;
    public final String nodeText;
    public final int childrenLength;
    // Identification Properties
    public final Xpath xpath;
    public final String cssSelector;
    // Position & Size
    public final int x;
    public final int y;
    public final int width;
    public final int height;
    // Attributes
    public final Map<String, String> attributes;
    public final Map<String, String> cssStyles;
    public String neighborText;

    public Element(
            String tag,
            String nodeText,
            String innerText,
            int childrenLength,
            Xpath xpath,
            String cssSelector,
            int x, int y,
            int width, int height,
            Map<String, String> attributes,
            Map<String, String> cssStyles
    ) {
        this.tag = tag;
        this.innerText = innerText;
        this.nodeText = nodeText;
        this.childrenLength = childrenLength;
        this.xpath = xpath;
        this.cssSelector = cssSelector;
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
        this.attributes = attributes;
        this.cssStyles = cssStyles;
    }

    public Element(Map<String, Object> map) {
        this.tag = (String) map.get("tag");
        this.innerText = (String) map.get("innerText");
        this.nodeText = (String) map.get("nodeText");
        this.childrenLength = (int) (long) map.get("childrenLength");
        this.xpath = new Xpath(
                (String) checkedCastToMap(map.get("xpath")).get("chrome"),
                (String) checkedCastToMap(map.get("xpath")).get("selenium"),
                (String) checkedCastToMap(map.get("xpath")).get("id")
        );
        this.cssSelector = (String) map.get("cssSelector");
        this.x = (int) (long) map.get("x");
        this.y = (int) (long) map.get("y");
        this.width = (int) (long) map.get("width");
        this.height = (int) (long) map.get("height");
        this.attributes = safeCastToMap(map.get("attributes"));
        this.cssStyles = safeCastToMap(map.get("cssStyles"));
    }

    private static Map<String, String> safeCastToMap(Object obj) {
        if (obj instanceof Map<?, ?> rawMap) {
            Map<String, String> stringMap = new HashMap<>();
            for (Map.Entry<?, ?> entry : rawMap.entrySet()) {
                if (!(entry.getKey() instanceof String) || !(entry.getValue() instanceof String)) {
                    throw new IllegalArgumentException("Map contains non-String key/value");
                }
                stringMap.put((String) entry.getKey(), (String) entry.getValue());
            }
            return stringMap;
        } else {
            throw new IllegalArgumentException("Object is not a Map<String, String>");
            // Or return a default value, e.g., Collections.emptyMap();
        }
    }

    private StringBuilder appendMapJson(StringBuilder sb, Map<String, String> map) {
        sb.append("{");
        for (Map.Entry<String, String> entry : map.entrySet()) {
            sb.append("\"")
                    .append(escape(entry.getKey()))
                    .append("\":\"")
                    .append(escape(entry.getValue()))
                    .append("\",");
        }
        if (!map.isEmpty()) {
            sb.deleteCharAt(sb.length() - 1);
        }
        sb.append("}");
        return sb;
    }

    private String escape(String s) {
        if (s == null) return null;
        return s.replace("\"", "\\\"");
    }

    public static class Xpath {
        public final String chrome;
        public final String selenium;
        public final String id;

        public Xpath(String chrome, String selenium, String id) {
            this.chrome = chrome;
            this.selenium = selenium;
            this.id = id;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }

            if (!(o instanceof Xpath xpath)) {
                return false;
            }

            return Objects.equals(chrome, xpath.chrome) &&
                    Objects.equals(selenium, xpath.selenium) &&
                    Objects.equals(id, xpath.id);
        }
    }

    public int area() {
        return width * height;
    }

    public int shape() {
        return height == 0 ? 0 : (width * 100) / height;
    }

    public String typeName() {
        return attributes.get("type");
    }

    public String className() {
        return attributes.get("class");
    }

    public String value() {
        return attributes.get("value");
    }

    public String placeholder() {
        return attributes.get("placeholder");
    }

    public Map<String, String> attributes() {
        return attributes;
    }

    public String tag() {
        return tag;
    }

    public String name() {
        return attributes.get("name");
    }

    public String id() {
        return attributes.get("id");
    }

    public String href() {
        return attributes.get("href");
    }

    public String alt() {
        return attributes.get("alt");
    }

    public String ariaLabel() {
        return attributes.get("aria-label");
    }

    public String xpathChrome() {
        return xpath.chrome;
    }

    public String xpathId() {
        return xpath.id;
    }

    public int[] location() {
        return new int[]{x, y};
    }

    public int[] dimensions() {
        return new int[]{width, height};
    }

    public String neighborText() {
        return neighborText;
    }

    public boolean isButton() {
        if (tag == null) return false;
        if ("a".equals(tag) && Objects.requireNonNullElse(className(), "").contains("btn")) return true;
        if ("button".equals(tag)) return true;
        return "input".equals(tag) && ("button".equals(typeName()) ||
                "submit".equals(typeName()) || "reset".equals(typeName()));
    }

    public String visibleText() {
        return Stream.of(innerText, value(), placeholder())
                .filter(Objects::nonNull)
                .filter(s -> !s.trim().isEmpty())
                .findFirst().orElse(null);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }

        if (!(o instanceof Element element)) {
            return false;
        }

        return childrenLength == element.childrenLength &&
                x == element.x && y == element.y &&
                width == element.width && height == element.height &&
                Objects.equals(tag, element.tag) &&
                Objects.equals(innerText, element.innerText) &&
                Objects.equals(nodeText, element.nodeText) &&
                Objects.equals(xpath, element.xpath) &&
                Objects.equals(cssSelector, element.cssSelector) &&
                Objects.equals(attributes, element.attributes) &&
                Objects.equals(cssStyles, element.cssStyles);
    }

    @Override
    public int hashCode() {
        return Objects.hash(tag, nodeText, innerText, childrenLength, xpath,
                cssSelector, x, y, width, height, attributes, cssStyles);
    }

    public String json() {
        StringBuilder sb = new StringBuilder();
        sb.append("{");
        sb.append("\"tag\":\"").append(escape(tag)).append("\",");
        sb.append("\"innerText\":\"").append(escape(innerText)).append("\",");
        sb.append("\"nodeText\":\"").append(escape(nodeText)).append("\",");
        sb.append("\"xpath\":{");
        sb.append("\"chrome\":\"").append(escape(xpath.chrome)).append("\",");
        sb.append("\"selenium\":\"").append(escape(xpath.selenium)).append("\",");
        sb.append("\"id\":\"").append(escape(xpath.id)).append("\"");
        sb.append("},");
        sb.append("\"cssSelector\":\"").append(escape(cssSelector)).append("\",");
        sb.append("\"x\":").append(x).append(",");
        sb.append("\"y\":").append(y).append(",");
        sb.append("\"width\":").append(width).append(",");
        sb.append("\"height\":").append(height).append(",");
        sb.append("\"attributes\":");
        appendMapJson(sb, attributes).append(",");
        sb.append("\"cssStyles\":");
        appendMapJson(sb, cssStyles);
        sb.append("}");

        return sb.toString();
    }

    @Override
    public String xpath() {
        return xpathChrome();
    }
}
